STYLE_GUIDE.md

```diff
@@ -0,0 +1,190 @@
+
+## Various
+
+ * Reserve "Save" and "Load" for long-running persistence operations. When parsing bytes, use "Encode" or "Decode".
+ * Receiver variables should be at least 2 letters.  e.g. not `func (s \*Something) MyFunc(){}` but `func (sm *Something) MyFunc(){}`.  This makes refactoring easier when we rename Something to something else.
+ * Maintain consistency across the codebase.
+ * Do not use "instance" in function names.
+ * Functions that return functions should have the suffix `Fn`
+ * A struct generally shouldn’t have a field named after itself, aka. this shouldn't occur:
+``` golang
+type middleware struct {
+	middleware Middleware
+}
+```
+ * In comments, use "iff" to mean, "if and only if".
+ * Product names are capitalized, like "Tendermint", "Basecoin", "Protobuf", etc except in command lines: `tendermint --help`
+ * Acronyms are all capitalized, like "RPC", "gRPC", "API".  "MyID", rather than "MyId".
+ * Prefer errors.New() instead of fmt.Errorf() unless you're actually using the format feature with arguments. (otherwise it's needlessly inefficient, & it'll crap out on %*)
+
+## Importing Libraries
+
+Sometimes it's necessary to rename libraries to avoid naming collisions or ambiguity.
+
+ * Use [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports)
+ * Separate imports into blocks - one for the standard lib and one for others. In some cases, it's nice to
+ separate into three: standard lib, external libs, tendermint libs.
+ * Here are some common library labels for consistency:
+   - dbm "github.com/tendermint/tmlibs/db"
+   - cmn "github.com/tendermint/tmlibs/common"
+   - tmcmd "github.com/tendermint/tendermint/cmd/tendermint/commands"
+   - tmcfg "github.com/tendermint/tendermint/config/tendermint"
+   - tmtypes "github.com/tendermint/tendermint/types"
+ * Never use anonymous imports (the `.`), for example, `tmlibs/common` or anything else.
+ * tip: Use the `_` library import to import a library for initialization effects (side effects)
+
+## Dependencies
+
+ * Use [dep](https://github.com/golang/dep).
+ * Never edit the Gopkg.lock file, instead if you need to lock a dependancy to a certain git hash, the `version` can
+ be set to the desired hash in Gopkg.toml. For more information on advanced usage of the `Gopkg.toml` file see
+[this](https://golang.github.io/dep/docs/daily-dep.html).
+ * Dependencies should be pinned by a release tag, or specific commit, to avoid breaking `go get` when external dependencies are updated.
+
+## Testing
+
+ * The first rule of testing is: we add tests to our code
+ * The second rule of testing is: we add tests to our code
+ * For Golang testing:
+   * Make use of table driven testing where possible and not-cumbersome
+     - [Inspiration](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go)
+   * Make use of [assert](https://godoc.org/github.com/stretchr/testify/assert) and [require](https://godoc.org/github.com/stretchr/testify/require)
+ * For Bash testing:
```

Comment:
Let's drop this section. We shouldn't have any significant logic in Bash - if we do, it should be rewritten in a "real" language.